src/types/tdigest.h

```diff
@@ -150,3 +151,74 @@ inline StatusOr<double> TDigestQuantile(TD&& td, double q) {
   diff /= (lc.weight / 2 + rc.weight / 2);
   return Lerp(lc.mean, rc.mean, diff);
 }
+
+template <typename TD>
+inline Status TDigestRank(TD&& td, const std::vector<double>& inputs, std::vector<int>& result) {
+  std::vector<size_t> indices(inputs.size());
+  std::iota(indices.begin(), indices.end(), 0);
+  std::sort(indices.begin(), indices.end(), [&inputs](size_t a, size_t b) { return inputs[a] < inputs[b]; });
+
+  result.resize(inputs.size());
+  size_t i = indices.size();
+  double cumulative_weight = 0;
+
+  // handle inputs larger than maximum
+  while (i > 0 && inputs[indices[i - 1]] > td.Max()) {
+    result[indices[i - 1]] = -1;
+    i--;
+  }
+
+  // reverse iterate through centroids and calculate reverse rank for each input
+  auto iter = td.End();
+  while (i > 0) {
+    auto centroid = GET_OR_RET(iter->GetCentroid());
+
+    if (centroid.mean > inputs[indices[i - 1]]) {
+      // mean > input, accumulate weight and move to prev centroid
+      cumulative_weight += centroid.weight;
+    } else if (centroid.mean == inputs[indices[i - 1]]) {
+      // mean == input, calculate reverse rank with half weight of current centroid
+      cumulative_weight += centroid.weight;
+      auto current_mean = centroid.mean;
+      auto current_mean_cumulative_weight = cumulative_weight + centroid.weight / 2;
+
+      // handle all the prev centroids which has the same mean
+      while (!iter->IsAtBegin() && iter->Prev()) {
+        auto next_centroid = GET_OR_RET(iter->GetCentroid());
+        if (current_mean != next_centroid.mean) {
+          // move back to the last equal centroid, because we will process it in the next loop
+          iter->Next();
+          break;
+        }
+        current_mean_cumulative_weight += centroid.weight / 2;
+        cumulative_weight += centroid.weight;
+      }
+
+      // assign the reverse rank for the inputs[indices[i - 1]]
+      result[indices[i - 1]] = static_cast<int>(current_mean_cumulative_weight);
+      i--;
+
+      // handle the prev inputs which has the same value
+      while ((i > 0) && (inputs[indices[i]] == inputs[indices[i - 1]])) {
```

Comment:
Maybe we could use a `map<double, std::vector<size_t>>` to construct a sorted input and group redundant input with rank into a group, and it may simplify our logic.